### DISCUSSION TITLE
text content did not match server-rendered html

### DISCUSSION THREAD

**Discussion by pacorroX** [vercel/next.js#51804] [Help] [answered]
### Summary

Hello, why when I create a Next.js app using the App Router and want to have my page.js as a Server Component but when I put a Client Component and make a log it logs on the Server and on the Client? Isn't suposed to work only in the client since it has the 'use client' at the top of the component?

I also get this error:
![image](https://github.com/vercel/next.js/assets/42702133/6c21cba3-5dec-40e1-9a9f-7d5307fe0c4f)

page.js
```
import Component from "@/app/components/Component";

export default function Home() {
    return (
        <main>
            <p>hello</p>
            <Component/>
        </main>
    )
}
```

Component.js
```
'use client'

import {useState} from "react";

export default function Component() {
    const [state, setState] = useState(new Date().getTime())

    console.log(state)

    return (<p>{state}</p>)
}
```

What is going on and what can be done to solve this and any future implementation of something like this?

I wanna have the page.js as a Server Component and the Component as a Client Component, I thought that was the approach with all of this.

Thanks.

### Additional information

```js
This can be recreated by using create-next-app and changing just the page.js and adding that Component.
```

**Comment by rasif-sahl** [comment, marked as answer]
In your code, it seems you are using the comment 'use client' in your Component.js file. However, the comment itself does not have any special meaning in Next.js or React. The behavior you're experiencing is expected because both the server and the client execute the code inside the component.

When you render a page in Next.js, the server-side code generates the initial HTML that is sent to the client. This includes executing the server-side code of the components, including the code inside your Component.js file. Once the initial HTML is sent to the client, the client-side code takes over and rehydrates the React components, running the component code again on the client side.

```
// page.js
import dynamic from 'next/dynamic';

const Component = dynamic(() => import('@/app/components/Component'), { ssr: false });

export default function Home() {
  return (
    <main>
      <p>hello</p>
      <Component />
    </main>
  );
}
```

```
// Component.js
import { useState, useEffect } from 'react';

export default function Component() {
  const [state, setState] = useState('');

  useEffect(() => {
    setState(new Date().getTime());
  }, []);

  console.log(state); // This log will only appear in the client console

  return <p>{state}</p>;
}
```
In the updated code, the dynamic function from next/dynamic is used to dynamically import the Component module. By setting { ssr: false }, you instruct Next.js not to perform server-side rendering for this component and ensure that it only runs on the client. This way, the server will only execute the server-side code, and the client will execute the client-side code.

**Comment by pacorroX** [comment]
Okay, now it works as I expected. And has this solution some tricky problems or can I use it normally?

And is this behavior new? I've never had this problem with pages/

**Comment by icyJoseph** [comment]
Hi,

The `use client` directive, does't mean that the component runs client side only. That's a widely distributed misconception!

Client Components, run both client and server side. Server Components are the new thing, and they run only server side.

- https://github.com/reactwg/server-components/discussions/4

If you really just need to use the timestamp, you can use the https://legacy.reactjs.org/docs/dom-elements.html#suppresshydrationwarning 

> However, in some rare cases, it is very hard or impossible to guarantee an exact match. For example, timestamps are expected to differ on the server and on the client.

**Comment by pacorroX** [comment]
Okay thanks!
